### ISSUE TITLE
make_dvcyaml: always save dvc.yaml

### ISSUE THREAD

**Issue by dberenbaum** [iterative/dvclive#381]
After playing around with `save_dvc_exp` a bit, I think it should always call `make_dvcyaml` even when run with `dvc exp run` or `dvc repro`. It makes the transition to using these smoother since you can keep the existing `dvc.yaml` to configure the metrics and plots.

Edit: I think because my `dvclive/dvc.yaml` file was untracked, it wasn't getting saved once I turned off `save_dvc_exp`. Regardless, I think we should overwrite it on each run to ensure it exists and is up to date.

**Comment by dberenbaum** [comment]
Some notes:

* If there is a `dvclive/dvc.yaml` but no `dvclive/params.yaml`, `dvc.api.params_show()` will complain. Can dvclive add/drop the params section in `dvclive/dvc.yaml` depending on whether `dvclive/params.yaml` exists?
* If I already have a root `dvc.yaml` with stage metrics and plots outputs and generate `dvclive/dvc.yaml`, it will duplicate plots but won't break anything.

If we can make params work, it might make sense to write out `dvclive/dvc.yaml` even when `save_dvc_exp=False`.

**Comment by dberenbaum** [comment]
Another scenario to consider: if I run `dvc exp run` with checkpoints but without metrics/plots configured, it would be useful to call `make_dvcyaml` at each step to get live metrics/plots.

**Comment by dberenbaum** [comment]
I plan to play around with this.

**Comment by dberenbaum** [comment]
@daavoo Added two possibilities in #384 and #385. I'm happy to go with either option for now.

Both always write out `dvc.yaml`. I don't see much downside here. Even if dvc isn't installed, it means I could go back and see metrics, plots, and params if I later install it and add it to my repo.

Both also write out a new `dvc.yaml` at least for each live instance, since it's likely the format will change over time, and this ensures the updates get propagated to the file.

**Comment by dberenbaum** [comment]
Reopening because this behavior was deemed too aggressive in some cases. For now, we can revert this behavior.

Comment from @shcheklein:

> we are generating evaluation:dvc.yaml even when there is a dvc.yaml already in the root. This causes strange effects (rendering of extra plots, some of the one single value (for metrics, since we dump them into plots as well). And at the end - everything is good. I think it’s too noisy / not expected

Let's please keep thinking about how this should behave. I'd still like to be able to generate this when using pipelines so that I don't have to manually configure the plots, metrics, and params myself.

**Comment by dberenbaum** [comment]
In VS Code, hiding these plots can be done in a single click by deselecting the dvclive directory in the plots section in the sidebar, so I don't find it that painful, but maybe there are ways we can better detect whether to either write that dvc.yaml or whether those plots should be shown.

I think it's important to write out dvc.yaml for pipelines because it greatly simplifies the default setup for using dvclive with pipelines. Currently, users need to set up something like this:

```
stages:
  eval:
    cmd: python eval.py
    metrics:
      - dvclive/metrics.json:
    plots:
      - dvclive/plots/metrics
      - dvclive/plots/images
      - dvclive/plots/sklearn/confusion_matrix.json:
          template: confusion
          x: actual
          y: predicted
```

If dvclive writes out its own dvc.yaml, then the pipeline config becomes:

```
stages:
  eval:
    cmd: python eval.py
    outs:
      - dvclive
```

**Comment by daavoo** [comment]
> For now, we can revert this behavior.

Opened #395 as a compromise. We can revert behavior if that's not good enough

**Comment by shcheklein** [comment]
I agree, I like the simplicity. We should respect the existing projects though and figure out how to deal with the untracked files + confirm with the team that we are fine complicating the semantics (pipeline being generated within a pipeline).

I think for an initial implementation, should we let DVCLive:

- detect that there are _any_ plots, metrics, params (people already can see the table / plots), etc and give up on saving exps, give up on generating additional config. This way we won't be breaking the existing repos. And after all our goal was onboarding, not simplifying the existing repos.
- it should be done w/o DVC api. I don't like that we keep depending on it. It's really heavy, it's a very strong assumption that it's installed this way ...

**Comment by dberenbaum** [comment]
> * detect that there are _any_ plots, metrics, params (people already can see the table / plots), etc and give up on saving exps, give up on generating additional config. This way we won't be breaking the existing repos. And after all our goal was onboarding, not simplifying the existing repos.

I think this or several other options are fine for a short term fix, but I'm not sure where this leaves us long term. If I add an evaluation stage using dvclive to a project with unrelated params/metrics/plots, I'm stuck. It's also not obvious (which is true for nearly all options so far).

To avoid breaking repos, I'd rather revert and turn it off completely by default. Then we could push for a 2.0 release in the not too distant future where we either:
1. Turn this on by default with an opt out (we could do the same for saving exp refs).
2. Find some better logic for auto detecting what to write to dvc.yaml.

**Comment by shcheklein** [comment]
> . If I add an evaluation stage using dvclive to a project with unrelated params/metrics/plots, I'm stuck. It's also not obvious (which is true for nearly all options so far).

yep, that's why generating a root `dvc.yaml` can be better? We can do it once, it's visible and can be further customized if needed.

If root file exists w/o plots we can add a section there.

> . If I add an evaluation stage using dvclive to a project with unrelated params/metrics/plots, I'm stuck. It's also not obvious (which is true for nearly all options so far).

yes, that works fine. And for new users we will provide a snippet that enables it.

**Comment by dberenbaum** [comment]
@shcheklein I think you quoted the wrong thing above? I guess you mean you are okay with this:

> To avoid breaking repos, I'd rather revert and turn it off completely by default.

Can you confirm and we can at least move forward in the short term?




> yep, that's why generating a root `dvc.yaml` can be better? We can do it once, it's visible and can be further customized if needed.
> 
> If root file exists w/o plots we can add a section there.

I like the approach from @daavoo above to open a PR to test it. I think it's easy to try out all these approaches for a bit and see what feels most helpful/least annoying.

**Comment by shcheklein** [comment]
> Can you confirm and we can at least move forward in the short term?

yes, sorry, I meant that I'm fine with `I'd rather revert and turn it off completely by default.`.

**Comment by dberenbaum** [comment]
Coming back to this. I worry current behavior makes it look broken because most people who try it are existing users who have a pipeline already, and they see nothing change when adding `save_dvc_exp`.

I think the simplest fix is another option like `Live(dvcyaml=True)`. Technically, this might require a 2.0 release if defaulting to `True` (and we probably want this default for onboarding). This is transparent and enables users to see the extra functionality but easily turn it off.

> yep, that's why generating a root `dvc.yaml` can be better? We can do it once, it's visible and can be further customized if needed.

The main concern I have with this approach is that it assumes a static dvclive setup. I personally would make lots of changes over time and add more metrics, plots, etc., so writing out only once feels not that useful to me. Again, happy to try this out in a branch/PR and see if it seems better after playing around with it.

**Comment by shcheklein** [comment]
Okay, can we detect and skip generating it if there is an overlap in `to-be-generated .yaml` and the one that exists already? In all other cases keep generating. wdyt?

**Comment by dberenbaum** [comment]
> Okay, can we detect and skip generating it if there is an overlap in `to-be-generated .yaml` and the one that exists already? In all other cases keep generating. wdyt?

Do you mind giving explaining a bit more why you prefer it over `Live(dvcyaml=True)`?

I think we can do it, but I'm not sure it addresses my main concern (since DVCLive users with experiments presumably configured plots in their pipelines already):

> I worry current behavior makes it look broken because most people who try it are existing users who have a pipeline already, and they see nothing change when adding `save_dvc_exp`.

I'm also unsure how to decide what's overlapping and whether this will be easy for users to understand. When you ran into issues overlapping with https://github.com/iterative/vscode-dvc-demo/blob/main/dvc.yaml, those plot IDs are not the dvclive paths. We could assume if the path is used in any plot, we should consider it an overlap, but what if you actually want multiple plots that use the same file? For example, in vscode-dvc-demo, dvclive generates the confusion matrix but the same file gets used in the custom jitter plot.

**Comment by shcheklein** [comment]
Okay, before I answer, maybe I'm not fully understand what `dvcyaml=True` means. Could you clarify please? :)

>  For example, in vscode-dvc-demo, dvclive generates the confusion matrix but the same file gets used in the custom jitter plot.

can be an advanced case, where you would have to explicitly enable it then?

**Comment by dberenbaum** [comment]
> Okay, before I answer, maybe I'm not fully understand what `dvcyaml=True` means. Could you clarify please? :)

If `dvcyaml=True`, then the `dvc.yaml` gets written. If `dvcyaml=False`, it doesn't get written.

Maybe one difference in perspective is whether you see this `dvc.yaml` as a starter template for a project or whether you see it as part of the core configuration that stays with your project as it grows 🤔 ?

**Comment by shcheklein** [comment]
And it should be True by default (for simplicity)? And we would remove `save_dvc_exp`?

**Comment by dberenbaum** [comment]
> And it should be True by default (for simplicity)?

Yes.

> And we would remove `save_dvc_exp`?

I was thinking we would keep it as a separate option that only does what it says: save a dvc exp. That way, you can save exps but turn off writing dvc.yaml, or keep writing dvc.yaml to configure your dvclive outputs even when you are using `dvc exp run`.

**Comment by shcheklein** [comment]
Makes sense to me.

> Maybe one difference in perspective is whether you see this dvc.yaml as a starter template for a project or whether you see it as part of the core configuration that stays with your project as it grows 🤔 ?

I like this btw. I think it's a good point about making it useful even beyond onboarding.

**Comment by dberenbaum** [comment]
Should we go ahead and do a 2.0 release just to fix this issue?

**Comment by shcheklein** [comment]
Yep, I'm fine with this.

**Comment by dberenbaum** [comment]
Fixed in #443
